cyvlfeat/sift/phow.py

```diff
@@ -0,0 +1,132 @@
+from __future__ import division
+import math
+import numpy as np
+import matplotlib.pyplot as plt
+import matplotlib.image as mpimg
+import matplotlib.colors as colors
+import dsift
+
+
+def phow(image, verbose=False, fast=True, sizes=[4, 6, 8, 10], step=2, color='gray',
+         floatdescriptors=False, magnif=6, window_size=1.5, contrastthreshold=0.005):
+    # -------------------------------------------------------------------
+    #  Parse the arguments
+    # -------------------------------------------------------------------
+    dsiftOpts = ['norm', 'window_size', window_size]
+
+    if verbose:
+        dsiftOpts.append('verbose')
+
+    if fast:
+        dsiftOpts.append('fast')
+
+    if floatdescriptors:
+        dsiftOpts.append('floatdescriptors')
+
+    dsiftOpts.extend(['step', step])
+
+    # -------------------------------------------------------------------
+    # Extract the features
+    # -------------------------------------------------------------------
+    def rgb2gray(rgb):
+        return np.dot(rgb[..., :3], [0.299, 0.587, 0.114])
+
+    # standardize the image
+    imageSize = [image.shape[0], image.shape[1]]
+
+    color_lower = color.lower()
+    gray_lower = 'gray'.lower()
+    set1 = set(color_lower.split(''))
+    set2 = set(gray_lower.split(' '))
+    if (set1 == set2):
+        numChannels = 1
+
+    if image.shape[2] > 1:
+        image = mpimg.imread(image)
+        # print img
+        image = rgb2gray(image)
+        # print "\n printing grayscale \n"
+        # print gray
+
+    else:
+        numChannels = 3
+        if image.shape[2] == 1:
+            np.concatenate((image, image, image), axis=2)
+        elif color_lower == 'rgb':
+            pass
+        elif color_lower == 'opponent':
+            '''Note that the mean differs from the standard definition of opponent
+              space and is the regular intensity (for compatibility with
+              the contrast thresholding).
+              Note also that the mean is added pack to the other two
+              components with a small multipliers for monochromatic
+              regions.'''
+            mu = 0.3 * image[0, :, :] + 0.59 * image[1, :, :] + 0.11 * image[2, :, :]
```

Comment:
Assuming that the image is of shape `(H, W, C` where `H` is height, `W` is width and `C` is the number of channels, then `image[0, :, :]` selects the first row of the image which seems like the wrong thing to do since this looks like an RGB to grayscale conversion.